src/unstated.js

```diff
@@ -13,8 +13,20 @@ export class Container<State: {}> {
     this._listeners = [];
   }
 
-  setState(state: $Shape<State>) {
-    this.state = Object.assign({}, this.state, state);
+  setState(state: $Shape<State> | (State => $Shape<State>)): void {
+    const prevState = this.state;
+    const newState = typeof state === 'function' ? state(prevState) : state;
+
+    /**
+     * allow either early return or ternary return of previous state to
+     * propagate correctly through component tree by not calling _listeners
+     * and leaving old state in place.
+     */
+    if (newState == null || newState === prevState) {
+      this.state = prevState;
```

Comment:
Is this even necessary? Or should I just return?

Reply:
assignment indicates that it changes something, I think `return` would be more appropriate here